### ISSUE TITLE
Exception periodically reported via IDE alert as stacktrace.txt

### ISSUE THREAD

**Issue by slaiyer** [tonybaloney/pycharm-security#204]
**Describe the bug**
```
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_ARRAY but was BEGIN_OBJECT at line 2 column 15 path $.
	at com.google.gson.Gson.fromJson(Gson.java:944)
	at com.google.gson.Gson.fromJson(Gson.java:897)
	at security.packaging.SafetyDbChecker.load(SafetyDbChecker.kt:64)
	at security.packaging.SafetyDbChecker.<init>(SafetyDbChecker.kt:49)
	at security.packaging.PyPackageSecurityScan.checkPackages(PyPackageSecurityScan.kt:32)
	at security.packaging.PythonPackageVulnerabilityStartupTask$runActivity$1.run(PythonPackageVulnerabilityStartupTask.kt:16)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:962)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:472)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:235)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:235)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Expected BEGIN_ARRAY but was BEGIN_OBJECT at line 2 column 15 path $.
	at com.google.gson.stream.JsonReader.beginArray(JsonReader.java:351)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:80)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:187)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	... 22 more
```
Also, exception reporting mechanism isn't implemented, according to Pycharm's tooltip for the report button.

**To Reproduce**
Code or Steps to reproduce the behavior:
1. On plugin load at IDE startup, occasionally repeats

**Expected behavior**
No exception.

**Screenshots**

**Additional context**
 - Plugin Version: 1.23.1
 - PyCharm Version (or IDE version): PyCharm 2020.3.5 (Community Edition)
Build #PC-203.7717.81, built on March 25, 2021
Runtime version: 11.0.10+8-b1145.96 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 10.16
 - Python Version: 3.7.10

**Comment by tonybaloney** [comment]
Thanks for reporting 
@Slaiyer 
Which of the configuration options have you selected for scanning the SafetyDB?

**Comment by Slaiyer** [comment]
Use SafteyDB from GitHub URL

**Comment by ii343hbka** [comment]
same for me

**Comment by andylamp** [comment]
I am having the same issue, if the error pops up I have to restart the IDE in order to get things properly working again. This only happened when I updated to the 2021.1 branches of PyCharm. It's also fairly deterministic, in the sense that it happens quite often. I can share a stack trace if needed.

**Comment by tonybaloney** [comment]
The short version is, the GSON parser library I'm using to Parse JSON of the SafetyDB doesn't like the non-standard "metadata" tag that SafetyDB recently started adding on their GitHub copy.

IF you switch to a local copy and remove this, it will work. 

I could not find a solution to configuring GSON to ignore this invalid part of the JSON file

**Comment by andylamp** [comment]
I see, that does not sound very promising - is there an alternative that I can use for automatically upgrading the db? or all the ones use the weird formatting you describe?

**Comment by Slaiyer** [comment]
> The short version is, the GSON parser library I'm using to Parse JSON of the SafetyDB doesn't like the non-standard "metadata" tag that SafetyDB recently started adding on their GitHub copy.

Do you mean https://github.com/pyupio/safety-db/blob/master/data/insecure.json#L2-L5 for example?

**Comment by tonybaloney** [comment]
> > The short version is, the GSON parser library I'm using to Parse JSON of the SafetyDB doesn't like the non-standard "metadata" tag that SafetyDB recently started adding on their GitHub copy.
> 
> Do you mean https://github.com/pyupio/safety-db/blob/master/data/insecure.json#L2-L5 for example?

Yes, `"$metadata"` is used as the key but it has a totally different structure to all the other attributes. 

My other solution was to just remove lines 2-5 before parsing it. It's hacky- but it would get the job done

**Comment by andylamp** [comment]
@tonybaloney, I vote for the hack to get this working - unless of course there is another, more favourable, option.

**Comment by Slaiyer** [comment]
@tonybaloney parts of the following guide seem to address this issue particularly:
- https://www.baeldung.com/gson-exclude-fields-serialization#exclusionstrategy
- https://www.baeldung.com/gson-exclude-fields-serialization#1-with-classes-and-fields-names
- https://www.baeldung.com/gson-exclude-fields-serialization#4-extend-exclusion-strategy-to-deserialization

**Comment by Slaiyer** [comment]
Code references:

- https://github.com/eugenp/tutorials/blob/master/gson/src/test/java/com/baeldung/gson/serializationwithexclusions/SerializationWithExclusionsUnitTest.java#L34-L59
- https://github.com/eugenp/tutorials/blob/master/gson/src/test/java/com/baeldung/gson/serializationwithexclusions/SerializationWithExclusionsUnitTest.java#L62-L79

You may wish to use `addDeserializationExclusionStrategy(strategy)` instead of `addSerializationExclusionStrategy(strategy)` or `setExclusionStrategies(strategy)`.

I would open a PR, but my currrent setup is too restrictive to install JVM tooling and test these changes. Hope this helps!

**Comment by tonybaloney** [comment]
I just tried going down that rabbit hole. It doesn't seem that a deserializationExclusionStrategy will do the job because it can either  ignore a class or a field, designed for having certain attributes of a class you want to ignore. In this case, we're deserialising a `Map<K, V>`, so I need to either be less specific (if that is possible) of `V` to allow both Object and List, or somehow get the deserialiser to ignore it another way.
I also investigated writing and registering a Type Adapter, but that didn't seem to work either 😢 

Leaning toward the hack. Would need to buffer out the stream, remove the meta and then write it back into a stream for it to work.

Alternatively, I could just fork the safety-db file and modify it on my copy
